File: sdks/python/memsy/control.py
from __future__ import annotations

import time
from typing import Any

import httpx

from memsy._http import DEFAULT_MAX_RETRIES, DEFAULT_RETRY_BACKOFF, HttpCoreMixin
from memsy.control_resources.billing import BillingResource
from memsy.control_resources.events import EventsResource
from memsy.control_resources.interest import InterestResource
from memsy.control_resources.keys import KeysResource
from memsy.control_resources.usage import UsageResource
from memsy.exceptions import MemsyAPIError, MemsyConnectionError
from memsy.models import HealthResponse, MeResponse, RateLimitInfo, UsageInfo


class MemsyControlClient(HttpCoreMixin):
    """
    Synchronous client for the Memsy control-plane API (api/).

    Handles account management, billing, API key lifecycle, usage reporting,
    and event browsing. Separate from ``MemsyClient`` because the control-plane
    is a distinct service with its own base URL.

    Usage::

        control = MemsyControlClient(
            base_url="https://api.memsy.io/api",
            api_key="msy_...",
        )

        me = control.me()
        events = control.events.list(limit=20)

        # Admin-only (returns AuthorizationError for non-admin API keys)
        summary = control.usage.summary()
        invoices = control.billing.invoices()
        key = control.keys.create("ci-key", scopes=["read"])

    Sub-resource accessors::

        control.usage       — UsageResource
        control.billing     — BillingResource
        control.keys        — KeysResource
        control.events      — EventsResource
        control.interest    — InterestResource
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float = 30.0,
        max_retries: int = DEFAULT_MAX_RETRIES,
        retry_backoff: float = DEFAULT_RETRY_BACKOFF,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._max_retries = max_retries
        self._retry_backoff = retry_backoff
        self._client = httpx.Client(
            base_url=self._base_url,
            timeout=timeout,
            headers={"Authorization": f"Bearer {api_key}"},
        )
        self.usage = UsageResource(self)
        self.billing = BillingResource(self)
        self.keys = KeysResource(self)
        self.events = EventsResource(self)
        self.interest = InterestResource(self)

    def __enter__(self) -> MemsyControlClient:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def close(self) -> None:
        """Close the underlying HTTP connection pool."""
        self._client.close()

    def _request(
        self, method: str, path: str, **kwargs: Any
    ) -> tuple[Any, UsageInfo | None, RateLimitInfo | None]:
        """Make HTTP request with retry logic for 429s."""
        for attempt in range(self._max_retries + 1):
            try:
                response = self._client.request(method, path, **kwargs)
            except httpx.ConnectError as e:
                raise MemsyConnectionError(
                    f"Could not connect to Memsy control-plane at {self._base_url}: {e}"
                ) from e
            except httpx.TimeoutException as e:
                raise MemsyConnectionError(f"Request to Memsy control-plane timed out: {e}") from e

            if response.status_code == 429 and attempt < self._max_retries:
                retry_after = response.headers.get("Retry-After")
                wait_time = (
                    float(retry_after) if retry_after else self._retry_backoff * (2**attempt)
                )
                time.sleep(wait_time)
                continue

            usage, rate_limit = self._parse_response_headers(response)

            if not response.is_success:
                raise self._classify_error(response)

            if response.status_code == 204 or not response.content:
                return None, usage, rate_limit

            return response.json(), usage, rate_limit

        raise MemsyAPIError("Max retries exceeded", status_code=429, detail="")

    def me(self) -> MeResponse:
        """Return identity information for the authenticated caller."""
        data, _, _ = self._request("GET", "/me")
        return MeResponse.from_dict(data)

    def health(self) -> HealthResponse:
        """Check if the Memsy control-plane is healthy."""
        data, _, _ = self._request("GET", "/health")
        return HealthResponse.from_dict(data)

File: sdks/node/src/control.ts
import { BaseHttpClient, type BaseClientOptions } from "./http.js";
import { type HealthResponse, type MeResponse, parseMeResponse } from "./models.js";
import { KeysResource } from "./control_resources/keys.js";
import { UsageResource } from "./control_resources/usage.js";
import { BillingResource } from "./control_resources/billing.js";
import { EventsResource } from "./control_resources/events.js";
import { InterestResource } from "./control_resources/interest.js";

export type MemsyControlClientOptions = BaseClientOptions;

/**
 * Memsy control-plane client (api/).
 *
 * Handles account management, billing, API key lifecycle, usage reporting,
 * and console event browsing. Separate from MemsyClient because the
 * control-plane is a distinct service with its own base URL — typically
 * `https://api.memsy.io/api` (vs. `/v1` for the hot path).
 *
 * Sub-resources mirror the Python SDK's MemsyControlClient:
 *   control.keys       — API key CRUD (admin-only)
 *   control.usage      — usage summary + timeseries (admin-only)
 *   control.billing    — billing summary + invoices (admin-only)
 *   control.events     — raw event browsing (seat-required)
 *   control.interest   — Pro plan interest signaling
 */
export class MemsyControlClient extends BaseHttpClient {
  readonly keys: KeysResource;
  readonly usage: UsageResource;
  readonly billing: BillingResource;
  readonly events: EventsResource;
  readonly interest: InterestResource;

  constructor(options: MemsyControlClientOptions) {
    super(options);
    this.keys = new KeysResource(this);
    this.usage = new UsageResource(this);
    this.billing = new BillingResource(this);
    this.events = new EventsResource(this);
    this.interest = new InterestResource(this);
  }

  async me(): Promise<MeResponse> {
    const { data } = await this.request<Record<string, unknown>>("GET", "/me");
    return parseMeResponse(data);
  }

  async health(): Promise<HealthResponse> {
    const { data, usage, rateLimit } = await this.request<{
      status: string;
      version?: string;
      billing_enabled?: boolean;
      components?: Record<string, string>;
    }>("GET", "/health");

    return {
      status: data.status,
      version: data.version ?? "",
      billingEnabled: data.billing_enabled ?? null,
      components: data.components ?? null,
      usage,
      rateLimit,
    };
  }
}
